vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+n_{IR} &\sim B(I, 1 - e^{-\gamma \cdot dt)}
+\end{align*}$$
+
+## Implementing the SIR model using [`odin.dust`](https://mrc-ide.github.io/odin.dust/)
+
+The above equations can straightforwardly be written out using the odin DSL:
+
+```r
+## Definition of the time-step and output as "time"
+dt <- user(1)
+initial(time) <- 0
+update(time) <- (step + 1) * dt
+
+## Core equations for transitions between compartments:
+update(S) <- S - n_SI
+update(I) <- I + n_SI - n_IR
+update(R) <- R + n_IR
+
+## Individual probabilities of transition:
+p_SI <- 1 - exp(-beta * I / N * dt) # S to I
+p_IR <- 1 - exp(-gamma * dt) # I to R
+
+## Draws from binomial distributions for numbers changing between
+## compartments:
+n_SI <- rbinom(S, p_SI)
+n_IR <- rbinom(I, p_IR)
+
+## Total population size
+N <- S + I + R
+
+## Initial states:
+initial(S) <- S_ini
+initial(I) <- I_ini
+initial(R) <- 0
+
+## User defined parameters - default in parentheses:
+S_ini <- user(1000)
+I_ini <- user(10)
+beta <- user(0.2)
+gamma <- user(0.1)
+```
+
+This is converted to a C++ dust model, and compiled into a library in a single step, using [`odin.dust`](https://mrc-ide.github.io/odin.dust/). Save the above code as a file named `sir.R`. File names must not contain special characters.
+
+```{r}
+library(odin.dust)
+gen_sir <- odin_dust("sir.txt")
+```
+
+## Adding age structure to the model
+
+Adding age structure to the model consists of the following steps, which turn variables into arrays:
+
+- Define the number of age categories as a user parameter `N_age`.
+- Add age structure to each compartment, by adding square brackets to the lvalues.
+- Modify the rvalues to use quantities from the appropriate compartment, by adding an `i` index to the rvalues. These will automatically be looped over.
+- Where an age compartment needs to be reduced into a single compartment/variable, use `sum` (or another array function from https://mrc-ide.github.io/odin/articles/functions.html as appropriate)
+- Define the dimensions of all arrays, for example by setting `dim(S) <- N_age`.
+
+This would simply give `N_age` independent processes equivalent to the first model, scaled by the size of the population in each age category. To actually make this useful, you likely want to add some interaction or transitions between the compartments. An example of this would be to add an age-specific contact matrix, demonstrated below, which defines a different force of infection $\lambda$ for each age group. This is calculated by $$\lambda_i = \frac{\beta}{N} \cdot \sum_{j=1}^{N_{\mathrm{age}}} I_j m_{ij}$$ In the odin code:
+
+```r
+m[, ] <- user() # age-structured contact matrix
+s_ij[, ] <- m[i, j] * I[i]
+lambda[] <- beta / N * sum(s_ij[i, ])
+```
+
+The probability of infection of a susceptible is then indexed by this force of infection:
+
+```r
+p_SI[] <- 1 - exp(-lambda[i] * dt)
+```
+
+Putting this all together, the age structured SIR model is as follows:
+
+```r
```

Comment:
This is not really used in the above? It looks like it would fit better in the odin vignette to me